File: boolean.py
Z = ("c", "0")
O = ("c", "1")


def mk(kind, items):
    flat = []
    for item in items:
        if item[0] == kind:
            for child in item[1]:
                flat.append(child)
        else:
            flat.append(item)
    if len(flat) == 0:
        if kind == "a":
            return O
        return Z
    if len(flat) == 1:
        return flat[0]
    return (kind, tuple(flat))


def kids(node, kind):
    if node[0] != kind:
        return [node]
    out = []
    for child in node[1]:
        if child[0] == kind:
            for item in kids(child, kind):
                out.append(item)
        else:
            out.append(child)
    return out


def same(a, b):
    return sig(a) == sig(b)


def sig(node):
    kind = node[0]
    if kind == "c":
        return node[1]
    if kind == "v":
        return node[1]
    if kind == "n":
        return "n(" + sig(node[1]) + ")"
    parts = []
    for child in kids(node, kind):
        parts.append(sig(child))
    parts.sort()
    if kind == "a":
        return "a(" + ",".join(parts) + ")"
    return "o(" + ",".join(parts) + ")"


def comp(a, b):
    return (a[0] == "n" and same(a[1], b)) or (b[0] == "n" and same(b[1], a))


def has(items, target):
    for item in items:
        if same(item, target):
            return True
    return False


def sub(a, b):
    for item in a:
        if not has(b, item):
            return False
    return True


def rem1(items, target):
    out = []
    done = False
    for item in items:
        if not done and same(item, target):
            done = True
        else:
            out.append(item)
    return out


def prio(node):
    if node[0] in ("c", "v"):
        return 4
    if node[0] == "n":
        return 3
    if node[0] == "a":
        return 2
    return 1


def simple_not(node):
    return node[0] in ("c", "v", "n")


def show(node, parent=0):
    kind = node[0]
    if kind in ("c", "v"):
        text = node[1]
    elif kind == "n":
        child = node[1]
        text = show(child, 0)
        if simple_not(child):
            text = text + ","
        else:
            text = "(" + text + "),"
    else:
        joiner = "."
        if kind == "o":
            joiner = " + "
        parts = []
        for child in kids(node, kind):
            part = show(child, 0)
            if prio(child) < prio(node):
                part = "(" + part + ")"
            parts.append(part)
        text = joiner.join(parts)
    if prio(node) < parent:
        return "(" + text + ")"
    return text


def short(node):
    return show(node).replace(" + ", "+")


def parse(text):
    toks = []
    i = 0
    while i < len(text):
        ch = text[i]
        if ch in " \t\r\n":
            i += 1
        elif ch in "(),+.":
            toks.append(ch)
            i += 1
        elif ch in "01":
            toks.append(ch)
            i += 1
        elif ("A" <= ch <= "Z") or ("a" <= ch <= "z") or ch == "_":
            j = i + 1
            while j < len(text):
                cj = text[j]
                if ("A" <= cj <= "Z") or ("a" <= cj <= "z") or ("0" <= cj <= "9") or cj == "_":
                    j += 1
                else:
                    break
            toks.append(text[i:j])
            i = j
        else:
            raise ValueError("Unexpected character: " + ch)

    p = 0

    def cur():
        if p >= len(toks):
            return ""
        return toks[p]

    def eat(x):
        nonlocal p
        if cur() != x:
            raise ValueError("Expected '" + x + "' but found '" + cur() + "'.")
        p += 1

    def atom():
        nonlocal p
        t = cur()
        if t == "(":
            eat("(")
            node = expr()
            eat(")")
        elif t == "0":
            p += 1
            node = Z
        elif t == "1":
            p += 1
            node = O
        elif t and t[0] not in "(),+.":
            p += 1
            node = ("v", t)
        else:
            raise ValueError("Unexpected token: " + t)
        while cur() == ",":
            eat(",")
            node = ("n", node)
        return node

    def term():
        node = atom()
        parts = [node]
        while cur() == ".":
            eat(".")
            parts.append(atom())
        return mk("a", parts)

    def expr():
        node = term()
        parts = [node]
        while cur() == "+":
            eat("+")
            parts.append(term())
        return mk("o", parts)

    node = expr()
    if cur():
        raise ValueError("Unexpected token: " + cur())
    return node


def direct(node):
    kind = node[0]
    if kind in ("c", "v"):
        return False
    if kind == "n":
        child = node[1]
        return child[0] in ("c", "n", "a", "o")
    parts = kids(node, kind)
    if kind == "o":
        for item in parts:
            if item == O:
                return True
        if len(parts) > 1:
            for item in parts:
                if item == Z:
                    return True
        i = 0
        while i < len(parts):
            j = i + 1
            while j < len(parts):
                if same(parts[i], parts[j]) or comp(parts[i], parts[j]):
                    return True
                j += 1
            i += 1
        i = 0
        while i < len(parts):
            j = 0
            small = kids(parts[i], "a")
            while j < len(parts):
                if i != j and len(kids(parts[j], "a")) > len(small) and sub(small, kids(parts[j], "a")):
                    return True
                j += 1
            i += 1
        return False
    for item in parts:
        if item == Z:
            return True
    if len(parts) > 1:
        for item in parts:
            if item == O:
                return True
    i = 0
    while i < len(parts):
        j = i + 1
        while j < len(parts):
            if same(parts[i], parts[j]) or comp(parts[i], parts[j]):
                return True
            j += 1
        i += 1
    i = 0
    while i < len(parts):
        j = 0
        small = kids(parts[i], "o")
        while j < len(parts):
            if i != j and len(kids(parts[j], "o")) > len(small) and sub(small, kids(parts[j], "o")):
                return True
            j += 1
        i += 1
    return False


def step(node):
    kind = node[0]
    if kind in ("c", "v"):
        return None

    if kind == "n":
        hit = step(node[1])
        if hit:
            return (("n", hit[0]), hit[1])
        child = node[1]
        if child == Z or child == O:
            return (O if child == Z else Z, "Common sense")
        if child[0] == "n":
            return (child[1], "Double complement")
        if child[0] == "a":
            out = []
            for item in kids(child, "a"):
                out.append(("n", item))
            return (mk("o", out), "De Morgan's law")
        if child[0] == "o":
            out = []
            for item in kids(child, "o"):
                out.append(("n", item))
            return (mk("a", out), "De Morgan's law")
        return None

    parts = kids(node, kind)
    i = 0
    while i < len(parts):
        hit = step(parts[i])
        if hit:
            new = parts[:]
            new[i] = hit[0]
            return (mk(kind, new), hit[1])
        i += 1

    if kind == "o":
        for item in parts:
            if item == O:
                return (O, "Common sense")
        i = 0
        while i < len(parts):
            if parts[i] == Z and len(parts) > 1:
                return (mk("o", parts[:i] + parts[i + 1:]), "Common sense")
            i += 1
    else:
        for item in parts:
            if item == Z:
                return (Z, "Common sense")
        i = 0
        while i < len(parts):
            if parts[i] == O and len(parts) > 1:
                return (mk("a", parts[:i] + parts[i + 1:]), "Common sense")
            i += 1

    i = 0
    while i < len(parts):
        j = i + 1
        while j < len(parts):
            if same(parts[i], parts[j]):
                return (mk(kind, parts[:j] + parts[j + 1:]), "Tautology")
            if comp(parts[i], parts[j]):
                if len(parts) == 2:
                    return ((O if kind == "o" else Z), "Tautology")
                new = parts[:]
                new[i] = O if kind == "o" else Z
                del new[j]
                return (mk(kind, new), "Tautology")
            j += 1
        i += 1

    if kind == "o":
        i = 0
        while i < len(parts):
            small = kids(parts[i], "a")
            j = 0
            while j < len(parts):
                if i != j:
                    big = kids(parts[j], "a")
                    if len(big) > len(small) and sub(small, big):
                        return (mk("o", parts[:j] + parts[j + 1:]), "Absorption")
                j += 1
            i += 1
    else:
        i = 0
        while i < len(parts):
            small = kids(parts[i], "o")
            j = 0
            while j < len(parts):
                if i != j:
                    big = kids(parts[j], "o")
                    if len(big) > len(small) and sub(small, big):
                        return (mk("a", parts[:j] + parts[j + 1:]), "Absorption")
                j += 1
            i += 1

    i = 0
    while i < len(parts):
        left = kids(parts[i], "a" if kind == "o" else "o")
        j = i + 1
        while j < len(parts):
            right = kids(parts[j], "a" if kind == "o" else "o")
            common = []
            for item in left:
                if has(right, item) and not has(common, item):
                    common.append(item)
            if common:
                a = left[:]
                b = right[:]
                for item in common:
                    a = rem1(a, item)
                    b = rem1(b, item)
                if a and b:
                    if kind == "o":
                        rep = mk("a", [mk("a", common), mk("o", [mk("a", a), mk("a", b)])])
                    else:
                        rep = mk("o", [mk("o", common), mk("a", [mk("o", a), mk("o", b)])])
                    new = []
                    k = 0
                    while k < len(parts):
                        if k == i:
                            new.append(rep)
                        elif k != j:
                            new.append(parts[k])
                        k += 1
                    if len(short(mk(kind, new))) <= len(short(mk(kind, parts))):
                        return (mk(kind, new), "Distributive law")
            j += 1
        i += 1

    if kind == "a":
        i = 0
        while i < len(parts):
            if parts[i][0] == "o":
                rest = []
                j = 0
                while j < len(parts):
                    if j != i:
                        rest.append(parts[j])
                    j += 1
                out = []
                for item in kids(parts[i], "o"):
                    out.append(mk("a", [item] + rest))
                cand = mk("o", out)
                ok = direct(cand)
                if not ok:
                    for item in out:
                        if direct(item):
                            ok = True
                            break
                if ok:
                    return (cand, "Expansion of brackets")
            i += 1
    else:
        i = 0
        while i < len(parts):
            if parts[i][0] == "a":
                rest = []
                j = 0
                while j < len(parts):
                    if j != i:
                        rest.append(parts[j])
                    j += 1
                out = []
                for item in kids(parts[i], "a"):
                    out.append(mk("o", rest + [item]))
                cand = mk("a", out)
                ok = direct(cand)
                if not ok:
                    for item in out:
                        if direct(item):
                            ok = True
                            break
                if ok:
                    return (cand, "Expansion of brackets")
            i += 1

    return None


def to_nand(node):
    kind = node[0]
    if kind == "c":
        return node
    if kind == "v":
        return node
    if kind == "n":
        child = node[1]
        if child == Z or child == O:
            return node
        return to_nand(child)
    if kind == "a":
        parts = kids(node, "a")
        nanded = []
        for item in parts:
            nanded.append(("n", to_nand(item)))
        return mk("o", nanded)
    if kind == "o":
        parts = kids(node, "o")
        converted = []
        for item in parts:
            converted.append(to_nand(item))
        inner = mk("a", converted)
        return ("n", inner)
    return node


def to_nor(node):
    kind = node[0]
    if kind == "c":
        return node
    if kind == "v":
        return node
    if kind == "n":
        child = node[1]
        if child == Z or child == O:
            return node
        return to_nor(child)
    if kind == "o":
        parts = kids(node, "o")
        nored = []
        for item in parts:
            nored.append(("n", to_nor(item)))
        return mk("a", nored)
    if kind == "a":
        parts = kids(node, "a")
        converted = []
        for item in parts:
            converted.append(to_nor(item))
        inner = mk("o", converted)
        return ("n", inner)
    return node


def normalise(node):
    kind = node[0]
    if kind in ("c", "v"):
        return node
    if kind == "n":
        return ("n", normalise(node[1]))
    if kind == "a":
        return mk("a", [normalise(x) for x in kids(node, "a")])
    if kind == "o":
        return mk("o", [normalise(x) for x in kids(node, "o")])
    return node


print("  , for NOT")
print("  . for AND")
print("  + for OR")
print("Variables: single letters (A, B, ...) or multi-letter (AB, XYZ, ...)")
print("")
print("Mode:")
print("  1. Simplify")
print("  2. NAND form")
print("  3. NOR form")
mode = input("Choose: ").strip()
if mode == "":
    mode = "1"

print("")
s = input("Enter: ").strip()
if s == "":
    s = "((B,.A),.B,),+A.B"
    print("Using: " + s)

try:
    cur = parse(s)
except ValueError as err:
    print("Input error: " + str(err))
else:
    print("")
    print("Start: " + show(cur))
    n = 1
    while n <= 50:
        hit = step(cur)
        if not hit:
            break
        cur = hit[0]
        print(str(n) + ". " + show(cur) + "    (" + hit[1] + ")")
        n += 1
    print("Result: " + show(cur))

File: addin/src/modules/boolean/boolean.cpp
#include "boolean.hpp"

#include <algorithm>
#include <cctype>
#include <functional>
#include <stdexcept>

namespace casio::boolean
{
namespace
{

NodeRef make_const(char v)
{
    auto n = std::make_shared<Node>();
    n->kind = Kind::Const;
    n->text = std::string(1, v);
    return n;
}

NodeRef make_var(std::string name)
{
    auto n = std::make_shared<Node>();
    n->kind = Kind::Var;
    n->text = std::move(name);
    return n;
}

NodeRef make1(Kind k, NodeRef child)
{
    auto n = std::make_shared<Node>();
    n->kind = k;
    n->kids = {std::move(child)};
    return n;
}

NodeRef mk(Kind kind, std::vector<NodeRef> items)
{
    // Port of booleanProgram.mk(): flatten same-kind children.
    std::vector<NodeRef> flat;
    for(auto const &it : items) {
        if(it && it->kind == kind) {
            for(auto const &c : it->kids) flat.push_back(c);
        }
        else flat.push_back(it);
    }
    if(flat.empty()) {
        return (kind == Kind::And) ? O() : Z();
    }
    if(flat.size() == 1) return flat[0];
    auto n = std::make_shared<Node>();
    n->kind = kind;
    n->kids = std::move(flat);
    return n;
}

static void kids(NodeRef const &node, Kind kind, std::vector<NodeRef> &out)
{
    if(!node || node->kind != kind) {
        out.push_back(node);
        return;
    }
    for(auto const &ch : node->kids) {
        if(ch && ch->kind == kind) kids(ch, kind, out);
        else out.push_back(ch);
    }
}

static std::string sig(NodeRef const &node);

static bool has(std::vector<NodeRef> const &items, NodeRef const &target)
{
    for(auto const &it : items) if(sig(it) == sig(target)) return true;
    return false;
}

static bool sub(std::vector<NodeRef> const &a, std::vector<NodeRef> const &b)
{
    for(auto const &it : a) if(!has(b, it)) return false;
    return true;
}

static std::vector<NodeRef> rem1(std::vector<NodeRef> items, NodeRef const &target)
{
    bool done = false;
    std::vector<NodeRef> out;
    for(auto const &it : items) {
        if(!done && sig(it) == sig(target)) done = true;
        else out.push_back(it);
    }
    return out;
}

static int prio(NodeRef const &node)
{
    if(!node) return 0;
    if(node->kind == Kind::Const || node->kind == Kind::Var) return 4;
    if(node->kind == Kind::Not) return 3;
    if(node->kind == Kind::And) return 2;
    return 1;
}

static bool simple_not(NodeRef const &node)
{
    if(!node) return false;
    return node->kind == Kind::Const || node->kind == Kind::Var || node->kind == Kind::Not;
}

static std::string sig(NodeRef const &node)
{
    if(!node) return "";
    if(node->kind == Kind::Const) return node->text;
    if(node->kind == Kind::Var) return node->text;
    if(node->kind == Kind::Not) return "n(" + sig(node->kids.at(0)) + ")";

    Kind kind = node->kind;
    std::vector<std::string> parts;
    std::vector<NodeRef> flat;
    kids(node, kind, flat);
    for(auto const &ch : flat) parts.push_back(sig(ch));
    std::sort(parts.begin(), parts.end());
    std::string head = (kind == Kind::And) ? "a(" : "o(";
    std::string out = head;
    for(std::size_t i = 0; i < parts.size(); i++) {
        if(i) out += ",";
        out += parts[i];
    }
    out += ")";
    return out;
}

static std::vector<std::string> expand_vars(std::string token)
{
    // Port of booleanProgram.expand_vars().
    for(char &c : token) c = static_cast<char>(std::toupper(static_cast<unsigned char>(c)));
    if(token.size() <= 1) return {token};
    std::vector<std::string> result;
    std::string current;
    current.push_back(token[0]);
    for(std::size_t i = 1; i < token.size(); i++) {
        char ch = token[i];
        if(ch >= 'A' && ch <= 'Z') {
            result.push_back(current);
            current = std::string(1, ch);
        }
        else current.push_back(ch);
    }
    result.push_back(current);
    return result;
}

} // namespace

NodeRef Z()
{
    static NodeRef z = make_const('0');
    return z;
}

NodeRef O()
{
    static NodeRef o = make_const('1');
    return o;
}

bool same(NodeRef const &a, NodeRef const &b) { return sig(a) == sig(b); }

bool comp(NodeRef const &a, NodeRef const &b)
{
    // Port: return (a[0]=='n' and same(a[1], b)) or vice versa.
    if(!a || !b) return false;
    if(a->kind == Kind::Not && same(a->kids.at(0), b)) return true;
    if(b->kind == Kind::Not && same(b->kids.at(0), a)) return true;
    return false;
}

std::string show(NodeRef const &node)
{
    // Port of booleanProgram.show() (with caching removed).
    if(!node) return "";

    if(node->kind == Kind::Const || node->kind == Kind::Var) return node->text;

    if(node->kind == Kind::Not) {
        auto child = node->kids.at(0);
        std::string t = show(child);
        if(simple_not(child)) t += ",";
        else t = "(" + t + "),";
        if(prio(node) < 0) return "(" + t + ")";
        return t;
    }

    std::string joiner = ".";
    if(node->kind == Kind::Or) joiner = " + ";

    std::vector<NodeRef> flat;
    kids(node, node->kind, flat);
    std::vector<std::string> parts;
    for(auto const &ch : flat) {
        std::string part = show(ch);
        if(prio(ch) < prio(node)) part = "(" + part + ")";
        parts.push_back(part);
    }
    std::string text;
    for(std::size_t i = 0; i < parts.size(); i++) {
        if(i) text += joiner;
        text += parts[i];
    }
    if(prio(node) < 0) return "(" + text + ")";
    return text;
}

NodeRef parse(std::string const &text)
{
    // Port of booleanProgram.parse().
    std::string up;
    up.reserve(text.size());
    for(char c : text) up.push_back(static_cast<char>(std::toupper(static_cast<unsigned char>(c))));

    std::vector<std::string> toks;
    std::size_t i = 0;
    while(i < up.size()) {
        char ch = up[i];
        if(ch == ' ' || ch == '\t' || ch == '\r' || ch == '\n') {
            i++;
        }
        else if(std::string_view("(),+.").find(ch) != std::string_view::npos) {
            toks.emplace_back(1, ch);
            i++;
        }
        else if(ch == '0' || ch == '1') {
            toks.emplace_back(1, ch);
            i++;
        }
        else if((ch >= 'A' && ch <= 'Z') || ch == '_') {
            std::size_t j = i + 1;
            while(j < up.size()) {
                char cj = up[j];
                if((cj >= 'A' && cj <= 'Z') || (cj >= '0' && cj <= '9') || cj == '_') j++;
                else break;
            }
            auto expanded = expand_vars(up.substr(i, j - i));
            for(std::size_t k = 0; k + 1 < expanded.size(); k++) {
                toks.push_back(expanded[k]);
                toks.emplace_back(".");
            }
            toks.push_back(expanded.back());
            i = j;
        }
        else throw std::runtime_error(std::string("Unexpected character: ") + ch);
    }

    std::size_t p = 0;
    auto cur = [&]() -> std::string {
        if(p >= toks.size()) return "";
        return toks[p];
    };
    auto eat = [&](std::string const &x) {
        if(cur() != x) throw std::runtime_error("Expected '" + x + "' but found '" + cur() + "'.");
        p++;
    };

    std::function<NodeRef()> expr;
    std::function<NodeRef()> term;
    std::function<NodeRef()> atom;

    atom = [&]() -> NodeRef {
        std::string t = cur();
        NodeRef node;
        if(t == "(") {
            eat("(");
            node = expr();
            eat(")");
        }
        else if(t == "0") { p++; node = Z(); }
        else if(t == "1") { p++; node = O(); }
        else if(!t.empty() && t[0] != '(' && t[0] != ')' && t[0] != ',' && t[0] != '+' && t[0] != '.') {
            p++;
            node = make_var(t);
        }
        else throw std::runtime_error("Unexpected token: " + t);

        while(cur() == ",") {
            eat(",");
            node = make1(Kind::Not, node);
        }
        return node;
    };

    term = [&]() -> NodeRef {
        NodeRef node = atom();
        std::vector<NodeRef> parts{node};
        while(cur() == ".") {
            eat(".");
            parts.push_back(atom());
        }
        return mk(Kind::And, std::move(parts));
    };

    expr = [&]() -> NodeRef {
        NodeRef node = term();
        std::vector<NodeRef> parts{node};
        while(cur() == "+") {
            eat("+");
            parts.push_back(term());
        }
        return mk(Kind::Or, std::move(parts));
    };

    NodeRef node = expr();
    if(!cur().empty()) throw std::runtime_error("Unexpected token: " + cur());
    return node;
}

std::pair<NodeRef, std::string> step(NodeRef const &node)
{
    // Partial port: enough for early parity. Full port later.
    if(!node) return {nullptr, ""};
    Kind kind = node->kind;
    if(kind == Kind::Const || kind == Kind::Var) return {nullptr, ""};

    if(kind == Kind::Not) {
        auto hit = step(node->kids.at(0));
        if(hit.first) return {make1(Kind::Not, hit.first), hit.second};

        auto child = node->kids.at(0);
        if(same(child, Z()) || same(child, O())) {
            return {same(child, Z()) ? O() : Z(), "Common sense"};
        }
        if(child->kind == Kind::Not) return {child->kids.at(0), "Double complement"};
        if(child->kind == Kind::And) {
            std::vector<NodeRef> out;
            std::vector<NodeRef> parts;
            kids(child, Kind::And, parts);
            for(auto const &it : parts) out.push_back(make1(Kind::Not, it));
            return {mk(Kind::Or, std::move(out)), "De Morgan's law"};
        }
        if(child->kind == Kind::Or) {
            std::vector<NodeRef> out;
            std::vector<NodeRef> parts;
            kids(child, Kind::Or, parts);
            for(auto const &it : parts) out.push_back(make1(Kind::Not, it));
            return {mk(Kind::And, std::move(out)), "De Morgan's law"};
        }
        return {nullptr, ""};
    }

    // Recurse into children first.
    std::vector<NodeRef> parts;
    kids(node, kind, parts);
    for(std::size_t i = 0; i < parts.size(); i++) {
        auto hit = step(parts[i]);
        if(hit.first) {
            auto new_parts = parts;
            new_parts[i] = hit.first;
            return {mk(kind, std::move(new_parts)), hit.second};
        }
    }

    // Common sense short-circuits
    if(kind == Kind::Or) {
        for(auto const &it : parts) if(same(it, O())) return {O(), "Common sense"};
        if(parts.size() > 1) {
            for(std::size_t i = 0; i < parts.size(); i++) {
                if(same(parts[i], Z())) {
                    auto newp = parts;
                    newp.erase(newp.begin() + i);
                    return {mk(Kind::Or, std::move(newp)), "Common sense"};
                }
            }
        }
    }
    else { // And
        for(auto const &it : parts) if(same(it, Z())) return {Z(), "Common sense"};
        if(parts.size() > 1) {
            for(std::size_t i = 0; i < parts.size(); i++) {
                if(same(parts[i], O())) {
                    auto newp = parts;
                    newp.erase(newp.begin() + i);
                    return {mk(Kind::And, std::move(newp)), "Common sense"};
                }
            }
        }
    }

    // Tautology and complements
    for(std::size_t i = 0; i < parts.size(); i++) {
        for(std::size_t j = i + 1; j < parts.size(); j++) {
            if(same(parts[i], parts[j])) {
                auto newp = parts;
                newp.erase(newp.begin() + j);
                return {mk(kind, std::move(newp)), "Tautology"};
            }
            if(comp(parts[i], parts[j])) {
                if(parts.size() == 2) return {(kind == Kind::Or) ? O() : Z(), "Tautology"};
                auto newp = parts;
                newp[i] = (kind == Kind::Or) ? O() : Z();
                newp.erase(newp.begin() + j);
                return {mk(kind, std::move(newp)), "Tautology"};
            }
        }
    }

    return {nullptr, ""};
}

} // namespace casio::boolean

